exercises/react/test/test_react.c

```diff
@@ -0,0 +1,295 @@
+
+   cell_value_set(input, 2);
+   TEST_ASSERT_EQUAL_INT(0, cbinfo.times_called);
+
+   cell_value_set(input, 4);
+   TEST_ASSERT_EQUAL_INT(1, cbinfo.times_called);
+   TEST_ASSERT_EQUAL_INT(222, cbinfo.last_value);
+
+   free_reactor(r);
+}
+
+void test_callbacks_can_be_added_and_removed(void)
+{
+   struct reactor *r = create_reactor();
+   struct cell *input = create_input_cell(r, 11);
+   struct cell *output = create_compute1_cell(r, input, plus1);
+
+   struct cbinfo cbinfo1 = { -1, 0 };
+   callback_id cb1 = add_callback(output, &cbinfo1, cb_record_last_value);
+   struct cbinfo cbinfo2 = { -1, 0 };
+   add_callback(output, &cbinfo2, cb_record_last_value);
+
+   cell_value_set(input, 31);
+
+   remove_callback(output, cb1);
+   struct cbinfo cbinfo3 = { -1, 0 };
+   add_callback(output, &cbinfo3, cb_record_last_value);
+
+   cell_value_set(input, 41);
+
+   TEST_ASSERT_EQUAL_INT(1, cbinfo1.times_called);
+   TEST_ASSERT_EQUAL_INT(32, cbinfo1.last_value);
+   TEST_ASSERT_EQUAL_INT(2, cbinfo2.times_called);
+   TEST_ASSERT_EQUAL_INT(42, cbinfo2.last_value);
+   TEST_ASSERT_EQUAL_INT(1, cbinfo3.times_called);
+   TEST_ASSERT_EQUAL_INT(42, cbinfo3.last_value);
+
+   free_reactor(r);
+}
+
+void test_removing_a_callback_multiple_times_doesnt_interfere_with_others(void)
+{
+   struct reactor *r = create_reactor();
+   struct cell *input = create_input_cell(r, 11);
+   struct cell *output = create_compute1_cell(r, input, plus1);
+
+   struct cbinfo cbinfo1 = { -1, 0 };
+   callback_id cb1 = add_callback(output, &cbinfo1, cb_record_last_value);
+   struct cbinfo cbinfo2 = { -1, 0 };
+   add_callback(output, &cbinfo2, cb_record_last_value);
+   for (int i = 0; i < 10; ++i) {
+      remove_callback(output, cb1);
+   }
+
+   cell_value_set(input, 2);
+
+   TEST_ASSERT_EQUAL_INT(0, cbinfo1.times_called);
+   TEST_ASSERT_EQUAL_INT(1, cbinfo2.times_called);
+   TEST_ASSERT_EQUAL_INT(3, cbinfo2.last_value);
+
+   free_reactor(r);
+}
+
+static int minus1(int x)
+{
+   return x - 1;
+}
+
+static int times(int x, int y)
+{
+   return x * y;
+}
+
+void
+test_callbacks_are_only_called_once_even_if_multiple_dependencies_change(void)
+{
+   struct reactor *r = create_reactor();
+   struct cell *input = create_input_cell(r, 1);
+   struct cell *plus_one = create_compute1_cell(r, input, plus1);
+   struct cell *minus_one1 = create_compute1_cell(r, input, minus1);
+   struct cell *minus_one2 = create_compute1_cell(r, minus_one1, minus1);
+   struct cell *output = create_compute2_cell(r, plus_one, minus_one2, times);
+
+   struct cbinfo cbinfo = { -1, 0 };
+   add_callback(output, &cbinfo, cb_record_last_value);
+
+   cell_value_set(input, 4);
+
+   TEST_ASSERT_EQUAL_INT(1, cbinfo.times_called);
+   TEST_ASSERT_EQUAL_INT(10, cbinfo.last_value);
+
+   free_reactor(r);
+}
+
+static int minus(int x, int y)
+{
+   return x - y;
+}
+
+void
+test_callbacks_are_not_called_if_dependencies_change_but_output_value_doesnt_change
```

Comment:
This gets `indent.sh` in an oscillating loop.

Run it with

```c
void
test_callbacks_are_not_called_if_dependencies_change_but_output_value_doesnt_change
(void)
```

and `indent` will change the code to:


```c
void test_callbacks_are_not_called_if_dependencies_change_but_output_value_doesnt_change(void)
```

Run it again, and `indent` changes it back. So I will be forced to shorten the name.